File: otto-mobile/otto-mobile/TravelSurface.swift
import Combine
import CoreLocation
import Foundation
import MapboxMaps

enum TravelSurface: Equatable {
    case land
    case water
}

enum MapTravelSurfaceSampler {
    /// Boat-on-water chips on map presence markers. Disabled: each visible moving sharer polled
    /// Mapbox Tilequery every 2s per device (expensive; cosmetic only). Route Builder road snap is separate.
    static let waterSurfaceDetectionEnabled = false

    static let minSpeedMphForBoat = 2.0
    static let confirmationDuration: TimeInterval = 4.0
    static let sampleThrottle: TimeInterval = 2.0

    static func instantaneousSurface(speedMph: Double, onWater: Bool, onRoad: Bool) -> TravelSurface {
        guard speedMph >= minSpeedMphForBoat else { return .land }
        guard onWater, !onRoad else { return .land }
        return .water
    }

    static func classify(features: [QueriedRenderedFeature]) -> (onWater: Bool, onRoad: Bool) {
        var onWater = false
        var onRoad = false
        for feature in features {
            let sourceLayer = feature.queriedFeature.sourceLayer?.lowercased() ?? ""
            for layerId in feature.layers {
                let normalizedLayer = layerId.lowercased()
                if isWaterFeature(sourceLayer: sourceLayer, layerId: normalizedLayer) {
                    onWater = true
                }
                if isRoadFeature(sourceLayer: sourceLayer, layerId: normalizedLayer) {
                    onRoad = true
                }
            }
            if feature.layers.isEmpty {
                if isWaterFeature(sourceLayer: sourceLayer, layerId: "") {
                    onWater = true
                }
                if isRoadFeature(sourceLayer: sourceLayer, layerId: "") {
                    onRoad = true
                }
            }
        }
        return (onWater, onRoad)
    }

    private static func isWaterFeature(sourceLayer: String, layerId: String) -> Bool {
        isWaterLayer(sourceLayer) || isWaterLayer(layerId)
    }

    private static func isRoadFeature(sourceLayer: String, layerId: String) -> Bool {
        isRoadLayer(sourceLayer) || isRoadLayer(layerId)
    }

    static func isWaterLayer(_ layer: String) -> Bool {
        let tokens = ["water", "waterway", "lake", "ocean", "river"]
        return tokens.contains { layer.contains($0) }
    }

    static func isRoadLayer(_ layer: String) -> Bool {
        let tokens = ["road", "bridge", "highway", "street", "motorway", "trunk", "transportation"]
        return tokens.contains { layer.contains($0) }
    }

    static func sample(
        mapboxMap: MapboxMap,
        coordinate: CLLocationCoordinate2D,
        speedMph: Double,
        completion: @escaping (TravelSurface) -> Void
    ) {
        guard waterSurfaceDetectionEnabled else {
            completion(.land)
            return
        }
        // Mapbox Standard (`.standard`) does not expose basemap layers to `queryRenderedFeatures`.
        // Use Tilequery against streets data so water/road checks work at any zoom/style.
        _ = mapboxMap
        Task {
            let features = await MapboxTilequeryClient.fetchFeatures(at: coordinate) ?? []
            let classified = MapboxTilequeryClient.classifySurface(features: features)
            let instantaneous = MapTravelSurfaceSampler.instantaneousSurface(
                speedMph: speedMph,
                onWater: classified.onWater,
                onRoad: classified.onRoad
            )
            await MainActor.run {
                completion(instantaneous)
            }
        }
    }
}

@MainActor
final class TravelSurfaceHysteresis {
    private(set) var displayed: TravelSurface = .land
    private var pending: TravelSurface?
    private var pendingSince: Date?

    func reset() {
        displayed = .land
        pending = nil
        pendingSince = nil
    }

    func apply(instantaneous: TravelSurface, now: Date = Date()) {
        if instantaneous == displayed {
            pending = nil
            pendingSince = nil
            return
        }
        if pending == instantaneous {
            guard let since = pendingSince else { return }
            if now.timeIntervalSince(since) >= MapTravelSurfaceSampler.confirmationDuration {
                displayed = instantaneous
                pending = nil
                pendingSince = nil
            }
        } else {
            pending = instantaneous
            pendingSince = now
        }
    }
}

@MainActor
final class TravelSurfaceTracker: ObservableObject {
    @Published private(set) var surfacesByUserID: [String: TravelSurface] = [:]
    private var controllers: [String: TravelSurfaceHysteresis] = [:]
    private var lastSampleAt: [String: Date] = [:]

    func surface(for userID: String) -> TravelSurface {
        surfacesByUserID[userID] ?? .land
    }

    func removeUsers(notIn ids: Set<String>) {
        controllers = controllers.filter { ids.contains($0.key) }
        surfacesByUserID = surfacesByUserID.filter { ids.contains($0.key) }
        lastSampleAt = lastSampleAt.filter { ids.contains($0.key) }
    }

    func ingest(userID: String, instantaneous: TravelSurface, now: Date = Date()) {
        let controller: TravelSurfaceHysteresis
        if let existing = controllers[userID] {
            controller = existing
        } else {
            let created = TravelSurfaceHysteresis()
            controllers[userID] = created
            controller = created
        }
        controller.apply(instantaneous: instantaneous, now: now)
        surfacesByUserID[userID] = controller.displayed
    }

    func shouldSample(userID: String, now: Date = Date()) -> Bool {
        guard let last = lastSampleAt[userID] else { return true }
        return now.timeIntervalSince(last) >= MapTravelSurfaceSampler.sampleThrottle
    }

    func markSampled(userID: String, now: Date = Date()) {
        lastSampleAt[userID] = now
    }
}

File: otto-android/app/src/main/java/to/ottomot/driftd/map/TravelSurface.kt
package to.ottomot.driftd.map

import java.util.concurrent.TimeUnit

enum class TravelSurface {
    Land,
    Water,
}

object MapTravelSurfaceSampler {
    /** Boat-on-water map chips. Off: Tilequery polled every 2s per moving sharer per device (costly; cosmetic). */
    const val WATER_SURFACE_DETECTION_ENABLED = false

    const val MIN_SPEED_MPH_FOR_BOAT = 2.0
    private val CONFIRMATION_MS = TimeUnit.SECONDS.toMillis(4)
    private val SAMPLE_THROTTLE_MS = TimeUnit.SECONDS.toMillis(2)

    fun instantaneousSurface(
        speedMph: Double,
        onWater: Boolean,
        onRoad: Boolean,
    ): TravelSurface {
        if (speedMph < MIN_SPEED_MPH_FOR_BOAT) return TravelSurface.Land
        if (!onWater || onRoad) return TravelSurface.Land
        return TravelSurface.Water
    }

    fun isWaterLayer(layer: String): Boolean {
        val tokens = listOf("water", "waterway", "lake", "ocean", "river")
        return tokens.any { layer.contains(it) }
    }

    fun isRoadLayer(layer: String): Boolean {
        val tokens = listOf("road", "bridge", "highway", "street", "motorway", "trunk", "transportation")
        return tokens.any { layer.contains(it) }
    }

    fun classifyFeatureLayers(
        sourceLayer: String?,
        layerId: String?,
    ): Pair<Boolean, Boolean> {
        val source = sourceLayer.orEmpty().lowercase()
        val layer = layerId.orEmpty().lowercase()
        val onWater = isWaterLayer(source) || isWaterLayer(layer)
        val onRoad = isRoadLayer(source) || isRoadLayer(layer)
        return onWater to onRoad
    }

    fun sampleThrottleMs(): Long = SAMPLE_THROTTLE_MS

    fun confirmationMs(): Long = CONFIRMATION_MS
}

class TravelSurfaceHysteresis {
    var displayed: TravelSurface = TravelSurface.Land
        private set

    private var pending: TravelSurface? = null
    private var pendingSinceMs: Long? = null

    fun reset() {
        displayed = TravelSurface.Land
        pending = null
        pendingSinceMs = null
    }

    fun apply(
        instantaneous: TravelSurface,
        nowMs: Long = System.currentTimeMillis(),
    ) {
        if (instantaneous == displayed) {
            pending = null
            pendingSinceMs = null
            return
        }
        if (pending == instantaneous) {
            val since = pendingSinceMs ?: return
            if (nowMs - since >= MapTravelSurfaceSampler.confirmationMs()) {
                displayed = instantaneous
                pending = null
                pendingSinceMs = null
            }
        } else {
            pending = instantaneous
            pendingSinceMs = nowMs
        }
    }
}

class TravelSurfaceTracker {
    private val controllers = linkedMapOf<String, TravelSurfaceHysteresis>()
    private val lastSampleAtMs = linkedMapOf<String, Long>()
    private val surfacesByUserId = linkedMapOf<String, TravelSurface>()

    fun surfaceFor(userId: String): TravelSurface = surfacesByUserId[userId] ?: TravelSurface.Land

    fun snapshot(): Map<String, TravelSurface> = surfacesByUserId.toMap()

    fun removeUsersNotIn(ids: Set<String>) {
        controllers.keys.retainAll(ids)
        surfacesByUserId.keys.retainAll(ids)
        lastSampleAtMs.keys.retainAll(ids)
    }

    fun ingest(
        userId: String,
        instantaneous: TravelSurface,
        nowMs: Long = System.currentTimeMillis(),
    ) {
        val controller = controllers.getOrPut(userId) { TravelSurfaceHysteresis() }
        controller.apply(instantaneous, nowMs)
        surfacesByUserId[userId] = controller.displayed
    }

    fun shouldSample(
        userId: String,
        nowMs: Long = System.currentTimeMillis(),
    ): Boolean {
        val last = lastSampleAtMs[userId] ?: return true
        return nowMs - last >= MapTravelSurfaceSampler.sampleThrottleMs()
    }

    fun markSampled(
        userId: String,
        nowMs: Long = System.currentTimeMillis(),
    ) {
        lastSampleAtMs[userId] = nowMs
    }
}
